### PR TITLE
[RFC] E-Mail

### PR DESCRIPTION
**Pull request by jamiehynds**
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/ecs/blob/master/CONTRIBUTING.md)?
- For proposing substantial changes or additions to the schema, have you reviewed the [RFC process](https://github.com/elastic/ecs/blob/master/rfcs/README.md)?
- If submitting code/script changes, have you verified all tests pass locally using `make test`?
- If submitting schema/fields updates, have you generated new artifacts by running `make` and committed those changes?
- Is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- Have you added an entry to the [CHANGELOG.next.md](https://github.com/elastic/ecs/blob/master/CHANGELOG.next.md)?

[RFC Preview](https://github.com/jamiehynds/ecs/blob/jamiehynds-patch-2/rfcs/text/0008-email.md)

### CODE REVIEW

File: rfcs/text/0008-email.md
Comment (vpiserchia) [inline rfcs/text/0008-email.md:45]: should not be email.return.address?

File: rfcs/text/0008-email.md
Comment (vpiserchia) [inline rfcs/text/0008-email.md:32]: why not allowing multiple files as in an array ?

File: rfcs/text/0008-email.md
Comment (vpiserchia) [inline rfcs/text/0008-email.md:49]: email.to is already a field of type keyword, how can you define email.to as an object having domain as subfield? sorry for the silly question

File: rfcs/text/0008-email.md
Comment (vpiserchia) [inline rfcs/text/0008-email.md:42]: MTA?

**Review by vpiserchia — COMMENTED** [review]
A series of question

File: rfcs/text/0008-email.md
Comment (webmat) [inline rfcs/text/0008-email.md:48]: `email.to.address`

File: rfcs/text/0008-email.md
Comment (webmat) [inline rfcs/text/0008-email.md:139]:
```suggestion
* Stage 0: https://github.com/elastic/ecs/pull/999
```

File: rfcs/text/0008-email.md
Comment (jeffrysleddens) [inline rfcs/text/0008-email.md:48]: I think 'recipient' is a more common term when it comes to email than 'to'.
Also the schema should support multiple recipients.

File: rfcs/text/0008-email.md
Comment (BenB196) [reply rfcs/text/0008-email.md:45]: You have .address, but don't have .domain for email.reply_to and object.return, should be constant and have .domain as an additional field.

Comment (BenB196) [inline rfcs/text/0008-email.md:45]: .address exits for email.reply_to and object.return, but not .domain, should be constant and have .domain as an additional field.

File: rfcs/text/0008-email.md
Comment (jeffrysleddens) [inline rfcs/text/0008-email.md:38]: The schema should support and distinguish both the envelope/smtp 'from' and the header/mime 'from'.

File: rfcs/text/0008-email.md
Comment (vpiserchia) [reply rfcs/text/0008-email.md:48]: this really depends if you want to capture the semantics of the envelope/smtp and/or the email headers. This applies to other fields as well (to, cc, bcc, from)

File: rfcs/text/0008-email.md
Comment (BenB196) [inline rfcs/text/0008-email.md:87]: When looking at the current fields provided, one of my concerns is it appears that they don't fit well with the rest of ECS. I think this can be partially fixed with the use of aliases, though, I don't believe aliases are standard/common in ECS.

Examples:

email.from -> source
email.to|cc|bcc -> destination
email.file -> file
email.latency -> event.duration

File: rfcs/text/0008-email.md
Comment (vpiserchia) [reply rfcs/text/0008-email.md:32]: file already exists as ECS field, it would be great to reuse it

File: rfcs/text/0008-email.md
Comment (webmat) [reply rfcs/text/0008-email.md:48]: I agree there's nuances in recipients, and in how closely we want to represent each protocol.

After looking at a few raw emails, I'm starting to think we should keep the design of this field set pretty high level.

If capturing each protocol's nuances is desired or necessary, then I think we should consider working on a specific breakdown per protocol, as a separate step.

So I think the guiding principle we could use here is to capture the commonalities in `email.*`, even if it means merging some concepts, sometimes.

One point you're raising is pretty interesting, however. Should we capture each "type" of recipient in a different field, or in one parent field with an additional label to indicate which type of recipient this is?

The current proposal takes the former approach:

```
{ "email": {
  "to": [ {"address": "alice@example.com", ...} ],
  "cc": [ {"address": "bob@example.com", ...} ],
} }
```

The "recipient" suggestion would look like:

```
{ "email": {
  "recipients": [
    { "type": "to", "address": "alice@example.com", ...},
    { "type": "cc", "address": "bob@example.com", ...},
  ]
} }
```

🤔

**Review by webmat — COMMENTED** [review]
Excellent start, @jamiehynds.

Thanks everyone so far for chiming in. Looks like there's a lot of interest in getting support for email in ECS 😃 

I think there's a lot of straightforward suggestions already made (in my review and prior reviews) that we could apply to the current RFC.

There's also many big discussions that don't have an obvious answer. Those should be captured in the "Concerns" section, so that these problem statements are part of the RFC document itself, and that we carry to the next RFC stages. I've tried to point them out in my review.

File: rfcs/text/0008-email.md
Comment (vpiserchia) [reply rfcs/text/0008-email.md:41]: nevertheless the message_id captures the uniqueness of a mail.
I can see that different mail servers have specific ways of building this Message ID and could be interesting (for identification purposes) capturing such behaviour (and spot the anomalies). With this said, a multi-field mapping would make sense here:

```
 | `email.message_id` | keyword | Internet message ID of the message |
| `email.message_id.text` | text | Internet message ID of the message for full text search |
```

File: rfcs/text/0008-email.md
Comment (vpiserchia) [reply rfcs/text/0008-email.md:48]: I really like this one:

`{ "email": {
  "to": [ {"address": "alice@example.com", ...} ],
  "cc": [ {"address": "bob@example.com", ...} ],
} }`

this avoid the need for nested objects. Also open to a new one in the "related" field:

`related.emails: ["alice@example.com", "bob@example.com"]`

File: rfcs/text/0008-email.md
Comment (SHolzhauer) [reply rfcs/text/0008-email.md:42]: Why add a field for this, i.m.o this can be captured in `process.*` fields. Specifically the `process.name` field.

This would be in line with [the statement](https://github.com/elastic/ecs/pull/999/files#r512252692) on `email.action`

File: rfcs/text/0008-email.md
Comment (ebeahan) [inline rfcs/text/0008-email.md:18]: Another item that came to mind, and I think makes sense to capture for further discussion later:

Would `email` be worth considering as an additional allowed value for `event.category`?

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:7]: @webmat Would there be any reason to discuss having the "protocols" under email rather than a top field level?

We could for example threat smtp, imap and pop3 as we do with for example "user.*", where it is expected to be nested under email.* but could meet other usecases moving forward.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:18]: I would say +1 on this.

@ebeahan if we would create this new `event.category` we should also update which `event.type` we can combine with it and maybe see if we need new ones?

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:26]: I agree with this as well, we should reuse the ECS fields we already have when we can.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:30]: @webmat I think it would make more sense to add `related.email` and maybe also `related.domain` instead. It would conform more to the current structure of how we handle arrays like this without nesting too much.
What do you think?

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:38]: How would we distinguish which hash is for which file, and the same with size?

Looking at the comment above about changing file to attachments, we will still need to for example have a list of objects if we want to keep track about size, hash, extension belonging to a single file/attachment.
This kinda goes against the current features we normally support (lists have very limited possibilities in terms of visualizations and SIEM/Alert rules)

Are we thinking all of these fields should just be an array?

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:43]: +1 for this.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:42]: I agree with @SHolzhauer on this, I think the name of the process should follow the current ECS format.

When it comes to #2 from @webmat it is something that might be useful in a separate field.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:48]: From an ingest perspective and a visualization perspective I think its better to keep the "to, cc, bcc" field structure over type.
The point from @vpiserchia is a good one, because once we start having list of objects we will have issues with the internal structure when it comes to visualizations, parsing and aggregations.

File: rfcs/text/0008-email.md
Comment (ebeahan) [reply rfcs/text/0008-email.md:30]: Under the current guidance with the `related.*` fields, email addresses are captured in `related.user` and domains in `related.hosts`.

Rather we continue with this guidance or adjust and add additional `related.*` fields is a good conversation to have. 🤔

**Comment by P1llus** [comment]
@ebeahan @webmat 

Just added the first iteration of the changes we discussed, hopefully this commit should cover all the above comments for now, some which we could investigate for later.
I added one more to the list, which was `attachments_count`, outside of the `attachments` flattened fields. This will help for visualisations and parser writing.

Could you please let me know if my commit covered all of the topics we wanted to cover for now?

File: rfcs/text/0008-email.md
Comment (andrewstucki) [inline rfcs/text/0008-email.md:37]: What about renaming this to `return_path` -- it's a bit more descriptive of what I think you're actually going for here.

File: rfcs/text/0008-email.md
Comment (ebeahan) [inline rfcs/text/0008-email.md:47]:
```suggestion
| `event.duration` | The duration related to the email event. Could be the total duration in Quarantine, how long the email took to send from source to destination etc |
```

File: rfcs/text/0008-email.md
Comment (ebeahan) [inline rfcs/text/0008-email.md:118]: We've been using GitHub handles in the `People` section as a convention for RFCs. Here's an [example](https://github.com/elastic/ecs/blob/master/rfcs/text/0001-wildcard-data-type.md#people).

Can we do the same here as well?

File: rfcs/text/0008-email.md
Comment (ebeahan) [inline rfcs/text/0008-email.md:4]: Thanks for updating 👍

**Review by ebeahan — COMMENTED** [review]
Thanks @P1llus!

I think we should still list some of the discussed points in the `Concerns` section. We don't have to have resolutions at this point. We simply want to capture them now for further discussion later:

* The two items Mat captured in an earlier review thread: https://github.com/elastic/ecs/pull/999#discussion_r512244335
* Possibility of needing an email `event.category` and possibly new `event.type` values.
* Maybe worth noting that if the flattened data type is used for `email.*` fields, it would be the first time using this type in ECS?

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:37]: I agree on this point as well, will update it in the next commit.

**Comment by P1llus** [comment]
Adding in the latest changes @ebeahan :+1:

File: rfcs/text/0008-email.md
Comment (BenB196) [inline rfcs/text/0008-email.md:41]: This field really feels like it should be part of the related fields. Something like related.domains (though it currently doesn't exist, so it might be worth keeping here)

Comment (P1llus) [reply rfcs/text/0008-email.md:41]: Yeah this field is the outcome of a current discussion we have. Instead of having domain fields for bcc, cc, recipients etc, we decided currently to have them all as an array under one field. This might change in the upcoming stages. Thanks for the pointer, always happy to get feedback

File: rfcs/text/0008-email.md
Comment (BenB196) [inline rfcs/text/0008-email.md:34]: Some additional fields that might be useful for the emails.sender fields:
- email.sender.domain | keyword | Domain of the sender
- email.sender.registered_domain | keyword | Registered domain of the sender (this is useful when dealing with reporting that involves multiple sub-domains sending emails)

These fields are useful for pivoting across large amounts of emails, as well as in aggregations for reporting. (email.sender.domain currently falls under email.domains, but capturing direction in reports is not possible via a related field)

File: rfcs/text/0008-email.md
Comment (BenB196) [inline rfcs/text/0008-email.md:38]: I can see email.size being more useful as a long rather than a keyword. I don't see many use-cases where you need to search for a specific email size (where use of keyword would outweigh use of long), compared to use-cases for aggregating and reporting based on a numeric type.

File: rfcs/text/0008-email.md
Comment (BenB196) [reply rfcs/text/0008-email.md:41]: Yeah, my concern mainly with a related field is that you lose the directionality of the value. Which might be useful for some use-cases.

File: rfcs/text/0008-email.md
Comment (ebeahan) [reply rfcs/text/0008-email.md:34]: Thanks @BenB196 - Good point to note, and thanks for the feedback about the fields being useful for pivoting and aggs.

There was some work done recently (https://github.com/elastic/ecs/pull/981) to ensure the domain name breakdown fields (`*.domain` + `*.registered_domain`, `top_level_domain`, and `*.subdomain`) were consistent across field sets. I expect we'll want to apply that same uniformity here.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:41]: We do have the `email.direction` though, would that be sufficient? We would calculate the direction before moving the different domains into `email.domains` for example.

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:38]: That is a typo from my side, thanks for spotting it!

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:34]: I agree with both. This is something we had been discussing as well for email, though the discussion was currently around top_level_domain only.
Since sender is a single address, I do see the reasons why we should have maybe a separate domain field (outside of `email.domains`).

I will add in some additions based on the comment from @ebeahan as well, and we could use that as a basis for any upcoming discussions as it does not need to be set in stone just yet.

File: rfcs/text/0008-email.md
Comment (andrewstucki) [reply rfcs/text/0008-email.md]: Quick question about these comments. Is email authentication (i.e. SPF/DKIM/DMARC) in scope for this RFC? Mainly wondering because they're really a product of alignment on Return-Path (MAIL FROM), From, and DKIM-Signature mail headers. Not that I recommend all implementations that can capture these headers attempt to validate SPF/DKIM/DMARC, but it's totally possible given only the message content (and some fun DNS lookups).

I think that they fall under a pretty different realm than say spam/reputation scoring, analytics reports, MTA actions, or even email protocols that were mentioned.

File: rfcs/text/0008-email.md
Comment (dainperkins) [reply rfcs/text/0008-email.md:95]: latency might be redundant depending on the specific action being recorded, but I wouldn't equate email.to|from with source and destination (or client/server) network entities

File: rfcs/text/0008-email.md
Comment (dainperkins) [inline rfcs/text/0008-email.md:32]: What do we for see as values for this?, and from which address fields (to, cc, bcc) would it be categorized on?  
It seems to me like something that could potentially be difficult to implement, and not sure of the value for visualizations (but I could easily be missing something obvious, its been one of those days...)

File: rfcs/text/0008-email.md
Comment (webmat) [reply rfcs/text/0008-email.md:18]: Let's mention this in the RFC "Fields" section.

**Review by webmat — COMMENTED** [review]
This RFC has by far surpassed the criteria for stage 1. We have a new author (thanks for picking it up, Marius) and a sponsor. And the fields of the proposal are looking really good already.

Of course on such an interesting subject, I couldn't resist providing some detailed technical feedback below.

But I think we should merge this as a stage 1 after a few minor adjustments which I'll outline in the body of this review, and we should address the rest of the technical feedback in a stage 2 PR.

The only points I'd like to address before merging this at stage 1 are:

- [ ] adjusting the "People" heading
- [ ] making the list of people into a bullet list
- [ ] adjusting the "stage 1" link at the very bottom
- [ ] and populating the "references" section

Everything else in my review can be addressed in the stage 2 PR. If you have time to incorporate more feedback in your next session on this, all the better. But feel free to just address the points above so we can move forward.

File: rfcs/text/0008-email.md
Comment (webmat) [reply rfcs/text/0008-email.md:125]: Additional note. @jamiehynds discussed with @marc-gr and Marc will take over as the sponsor for this RFC. Could you mark him as such in the People section, please?

Comment (webmat) [reply rfcs/text/0008-email.md:125]: Sorry, disregard this comment about the sponsor. I misunderstood what Jamie said.

Jamie remains the sponsor for this one :-)

File: rfcs/text/0008-email.md
Comment (P1llus) [reply rfcs/text/0008-email.md:33]: I think its worth exploring in the upcoming stage for sure if that is appropriate.

**Comment by P1llus** [comment]
Added the required changes you mentioned + a few more tweaks based on some changes that could wait for stage 2 @webmat . Anything else we should do at this point?

**Review by webmat — APPROVED** [review]
Alright, thanks for the adjustments, @P1llus! I'll adjust the date and merge.

File: rfcs/text/0008-email.md
Comment (webmat) [inline rfcs/text/0008-email.md:5]:
```suggestion
- Date: **2020-11-30** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
```

**Pull request merged**
